include/seqan3/search/fm_index/fm_index.hpp

```diff
@@ -438,6 +452,34 @@ class fm_index
         return {*this};
     }
 
+    /*!\brief Swap contents with another instance.
+     * \param lhs The first instance.
+     * \param rhs The other instance to swap with.
+     *
+     * ###Complexity
+     *
+     * Constant.
+     *
+     * ###Exceptions
+     *
+     * No-throw guarantee.
+     */
+    friend void swap(fm_index & lhs, fm_index & rhs) noexcept
+    {
+        using std::swap;
+
+        swap(lhs.index, rhs.index);
+        swap(lhs.sigma, rhs.sigma);
+        swap(lhs.text_begin, rhs.text_begin);
+        swap(lhs.text_begin_ss, rhs.text_begin_ss);
+        swap(lhs.text_begin_rs, rhs.text_begin_rs);
+
+        lhs.text_begin_ss.set_vector(&lhs.text_begin);
+        lhs.text_begin_rs.set_vector(&lhs.text_begin);
+        rhs.text_begin_ss.set_vector(&rhs.text_begin);
+        rhs.text_begin_rs.set_vector(&rhs.text_begin);
+    }
+
```

Comment:
Since C++11 Copy'n'swap is typically implemented like this:

  * copy constructor with copy implementation
  * move constructor with move implementation
  * assignment operator that takes by value and std::swaps
  * You very rarely need to define a `.swap()` member since C++11. Always use std::swap which uses the move constructor automatically.

Reply:
I would write it like this:

```cpp
fm_index(fm_index const & rhs) :
    index{rhs.index}, sigma{rhs.sigma}, text_begin{rhs.text_begin}, text_begin_ss{rhs.text_begin_ss}
{
    text_begin_ss.set_vector(&text_begin);
    text_begin_rs.set_vector(&text_begin);
}

fm_index(fm_index && rhs) :
    index{std::move(rhs.index)}, sigma{std::move(rhs.sigma)}, text_begin{std::move(rhs.text_begin)}, text_begin_ss{std::move(rhs.text_begin_ss)}
{
    text_begin_ss.set_vector(&text_begin);
    text_begin_rs.set_vector(&text_begin);
}

fm_index & operator=(fm_index rhs) //!< Delegate to swap().
{
    std::swap(*this, rhs);
    return *this;
}
```

This clearly highlights that only thing non-standard that happens is calling `set_vector` for two members.

Reply:
Mhm, using `std::swap` segfaults. It doesn't seem to set the pointers correctly.

Reply:
Hm, that sounds weird. Then again, maybe std::swap uses assignment which uses std::swap ... :thinking: 

In any case, if it works now and `std::swap` works now as well, we could also just put move assignment of the member into the assignment operator and get rid of the swap member, or not? I really don't like having a swap member, it looks weird :innocent:

Reply:
> Hm, that sounds weird. Then again, maybe std::swap uses assignment which uses std::swap ... 🤔
> 

It kinda looked like it in the debugger.
> In any case, if it works now and `std::swap` works now as well, we could also just put move assignment of the members (no swap required) into the assignment operator and get rid of the swap member, or not? I really don't like having a swap member, it looks weird 😇
Ok, done!